zqd/api/connection.go

```diff
@@ -0,0 +1,364 @@
+package api
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path"
+	"strconv"
+	"strings"
+
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"time"
+
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio/bzngio"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+const (
+	// DefaultPort zqd port to connect with.
+	DefaultPort = 9867
+)
+
+var (
+	// ErrSpaceNotFound returns when specified space does not exist.
+	ErrSpaceNotFound = errors.New("space not found")
+	// ErrSpaceExists returns when specified the space already exists.
+	ErrSpaceExists = errors.New("space exists")
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type Connection struct {
```

Comment:
For the lower-level connection setup, transport options, and http roundtrip, would you look into using a request library, like resty ( https://github.com/go-resty/resty )? It supports the features we need now, and is tested & documented, so it should be easy for us to add support for other features (like authentication options). It also allows getting raw access to the response reader, so it will be possible to handle our ndjson streaming responses with it.

Reply:
I reimplemented this using resty as you suggested. Having gotten pretty familiar with the resty api by now here's my takeaway:

It works great for simple our simple request/response w/ json payloads. OTH on more complicated requests it more often than not seems to get in the way. For instance: For our streaming responses I have to disable response parsing to prevent resty from trying to read the entire response into memory. This has the side effect of disabling all middleware, which means I had to reimplement this functionality on streaming requests (i.e. search and packet post). I think we should still use resty, however, I imagine in the future we're going to want to implement our own version of resty that is easier to work with streaming data. The vault api client does something similar to this.

Reply:
Do you mean we would implement all of resty's functionality, or that we would try to add ndjson streaming support to our own fork of resty?

Reply:
I suppose I meant reimplementing the resty functionality that we use. But really this is tbd